### DISCUSSION TITLE
How to increase fingerprint emulation

### DISCUSSION THREAD

**Discussion by AlexPaiva** [seleniumbase/SeleniumBase#2394] [Q&A] [answered]
What other means of anti fingerprint can I implement that are randomized and emulated randomly when using UC that will help in making the fingerprint completely random when having a high number of threads on the same site but want each thread to look completely different from the first?

**Comment by mdmintz** [comment, marked as answer]
You could try using multiple proxy servers to vary things up.
You'll have to supply your own proxy config for that to work.

```python
from parameterized import parameterized
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__, "-n3")

class ProxyTests(BaseCase):
    @parameterized.expand(
        [
            ["host1:port1"],
            ["host2:port2"],
            ["host3:port3"],
        ]
    )
    def test_multiple_proxies(self, proxy_string):
        self.get_new_driver(
            undetectable=True, proxy=proxy_string, multi_proxy=True
        )
        self.driver.get("https://browserleaks.com/webrtc")
        self.sleep(30)
```

![multi_webrtc_checks](https://github.com/seleniumbase/SeleniumBase/assets/6788579/9292cd57-f364-49a5-9cd3-3e81235a68ef)

**Comment by AlexPaiva** [comment]
> You could try using multiple proxy servers to vary things up. You'll have to supply your own proxy config for that to work.
> 
> ```python
> from parameterized import parameterized
> from seleniumbase import BaseCase
> BaseCase.main(__name__, __file__, "-n3")
> 
> class ProxyTests(BaseCase):
>     @parameterized.expand(
>         [
>             ["host1:port1"],
>             ["host2:port2"],
>             ["host3:port3"],
>         ]
>     )
>     def test_multiple_proxies(self, proxy_string):
>         self.get_new_driver(
>             undetectable=True, proxy=proxy_string, multi_proxy=True
>         )
>         self.driver.get("https://browserleaks.com/webrtc")
>         self.sleep(30)
> ```
> 
> ![multi_webrtc_checks](https://private-user-images.githubusercontent.com/6788579/278879558-9292cd57-f364-49a5-9cd3-3e81235a68ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwMTkyMzgsIm5iZiI6MTcwNDAxODkzOCwicGF0aCI6Ii82Nzg4NTc5LzI3ODg3OTU1OC05MjkyY2Q1Ny1mMzY0LTQ5YTUtOWNkMy0zZTgxMjM1YTY4ZWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDIzMTIzMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMzFUMTAzNTM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjY4NTIyNmU3ODJjZjg4ZjBkMGMzYjg2OTUyZWNiZjc2YTU0NGNiNWZjODZmNzRmYWU4YmIyNTAxNDgyZDcxMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2UPuqwzhzYyv13JDKrAZYHsC8D-WA-d8jzd8g6XCpT0)

Confused, I meant how to emulate the fingerprint not related to location such as canvas etc

**Reply by mdmintz** [reply]
See https://github.com/seleniumbase/SeleniumBase/discussions/2248#discussioncomment-7501491

Even before UC Mode existed, there was an integration with `selenium-stealth` (https://github.com/seleniumbase/SeleniumBase/issues/942#issuecomment-880771150), which lets you "change" (fake) various properties:

You can integrate it directly into the custom ``setUp()`` method of your tests like this after you pip install it:

```python
from seleniumbase import BaseCase
from selenium_stealth import stealth

class BaseTestCase(BaseCase):
    def setUp(self):
        super().setUp()
        stealth(self.driver,
            languages=["en-US", "en"],
            vendor="Google Inc.",
            platform="Win32",
            webgl_vendor="Intel Inc.",
            renderer="Intel Iris OpenGL Engine",
            fix_hairline=True,
        )
```

Then have your test classes inherit `BaseTestCase` instead of `BaseCase`. (See [SeleniumBase/help_docs/syntax_formats.md#sb_sf_02](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md#sb_sf_02))

There's general information on StackOverflow about that here: https://stackoverflow.com/q/70265306/7058266

Additionally, you could use `driver.execute_cdp_cmd()` to make internal modifications to your Chrome settings.
You would need to be familiar with CDP internals: https://chromedevtools.github.io/devtools-protocol/tot/Network/.
For example, here's a SeleniumBase script that changes your GeoLocation: [SeleniumBase/examples/test_geolocation.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_geolocation.py)

**Reply by AlexPaiva** [reply]
Oh wow didn't know you had ss integrated, how can I call it from a raw driver?

Any ideas on adding a randomizer for ss of some kind? Like calling ss completely randomized?

**Reply by mdmintz** [reply]
Wrap your `driver` with `stealth`:

```python
from seleniumbase import Driver
from selenium_stealth import stealth

driver = Driver(uc=True)
stealth(driver,
    languages=["en-US", "en"],
    vendor="Google Inc.",
    platform="Win32",
    webgl_vendor="Intel Inc.",
    renderer="Intel Iris OpenGL Engine",
    fix_hairline=True,
)
driver.get("https://browserleaks.com/webrtc")
driver.sleep(10)
driver.quit()
```

Check with https://github.com/diprajpatra/selenium-stealth for randomization techniques.
That's a separate repo that SeleniumBase integrations with.
